internal/retry/request.go

```diff
@@ -0,0 +1,248 @@
+// otherwise it returns 1 minute.
+func DefaultBackoff(d time.Duration) time.Duration {
+	if d <= 0 {
+		return time.Minute
+	} else {
+		return d * 2
+	}
+}
+
+type Options struct {
+	maxRetries         int
+	initialDelay       time.Duration
+	backoff            BackoffFn
+	sleep              sleepFn
+	retryAfter         RetryAfterFn
+	retryStrategyFuncs []RetryStrategyFn
+}
+type Option interface {
+	Apply(*Options)
+}
+type optionFn func(*Options)
+
+func (o optionFn) Apply(opts *Options) {
+	o(opts)
+}
+
+func MaxRetries(max int) Option {
+	return optionFn(func(o *Options) {
+		o.maxRetries = max
+	})
+}
+
+func Backoff(bo BackoffFn) Option {
+	return optionFn(func(o *Options) {
+		o.backoff = bo
+	})
+}
+
+func InitialDelay(d time.Duration) Option {
+	return optionFn(func(o *Options) {
+		o.initialDelay = d
+	})
+}
+
+func RetryAfter(ra RetryAfterFn) Option {
+	return optionFn(func(o *Options) {
+		o.retryAfter = ra
+	})
+}
+
+func Strategy(rs RetryStrategyFn) Option {
+	return optionFn(func(o *Options) {
+		o.retryStrategyFuncs = append(o.retryStrategyFuncs, rs)
+	})
+}
+
+func MakeOptions(os ...Option) *Options {
+	opts := &Options{
+		maxRetries:   DefaultMaxRetries,
+		initialDelay: DefaultInitialDuration,
+		sleep:        time.Sleep,
+		backoff:      DefaultBackoff,
+	}
+	for _, o := range os {
+		o.Apply(opts)
+	}
+	return opts
+}
+
+type Request struct {
+	client   func(*http.Request) (*http.Response, error)
+	r        *http.Request
+	o        *Options
+	attempts int
+	done     bool
+	delay    time.Duration
+}
+
+func NewRequest(r *http.Request, client func(*http.Request) (*http.Response, error), o *Options) *Request {
+	return &Request{
+		client: client,
+		r:      r,
+		o:      o,
+	}
+}
+
+func (r *Request) Done() bool {
+	return r.done || r.attempts >= r.o.maxRetries
+}
+
+func (r *Request) onError(err error) (*http.Response, error) {
+	r.done = true
+	return nil, err
+}
+
+func (r *Request) onDone(resp *http.Response, err error) (*http.Response, error) {
+	r.done = true
+	return resp, err
+}
+
+func (r *Request) Do() (*http.Response, error) {
+	if r.Done() {
+		// Already done, so don't attempt any more.
+		return nil, ErrorNoMoreAttempts
+	}
+	if r.attempts > 0 {
+		// This is a retry!
+		if r.delay > 0 {
+			// Wait if we have a delay
+			r.o.sleep(r.delay)
+		}
+		// Update the delay
+		r.delay = r.o.backoff(r.delay)
+	}
+	// Bump the number of attempts
+	r.attempts++
+
+	// Issue the request
+	resp, err := r.client(r.r)
+	if err != nil {
+		// Return the response and the error if the client returned an error
+		// TODO: pass err to the strategy funcs.
+		return r.onDone(resp, err)
+	}
+
+	// Immediate success if the request is 2xx or 3xx
+	if http.StatusOK <= resp.StatusCode && resp.StatusCode < http.StatusBadRequest {
+		return r.onDone(resp, err)
+	}
+
+	if r.o.retryAfter != nil {
+		// Check if the Retry-After header is set
+		d := r.o.retryAfter(resp)
```

Comment:
just curious: Why do we get the user to provide this function?

Reply:
Mostly for logging reasons at the moment (the current location has logging support). I'll add a TODO there to move it here later when I'm more confident in its correctness.